### PR TITLE
Fix: consumer invoker cache set nil after the ZK connection is lost

### PR DESCRIPTION
**Pull request by jack15083**
**What this PR does**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: remoting/zookeeper/client.go [hunk 1]
```diff
@@ -259,14 +259,14 @@ func (z *ZookeeperClient) HandleZkEvent(session <-chan zk.Event) {
...
 				logger.Warnf("zk{addr:%s} state is StateDisconnected, so close the zk client{name:%s}.", z.ZkAddrs, z.name)
-				z.stop()
 				z.Lock()
 				conn := z.Conn
 				z.Conn = nil
 				z.Unlock()
 				if conn != nil {
 					conn.Close()
 				}
+				z.stop()
```
Comment (Patrick0308) [inline remoting/zookeeper/client.go:269]: Stop first is correctly.  And then close connection.

File: registry/zookeeper/registry.go [hunk 2]
```diff
@@ -226,14 +226,15 @@ func (r *zkRegistry) registerTempZookeeperNode(root string, node string) error {
...
 			// should delete the old node
-			logger.Info("Register temp node failed, try to delete the old and recreate  (root{%s}, node{%s}) , ignore!", root, node)
+			/*logger.Info("Register temp node failed, try to delete the old and recreate  (root{%s}, node{%s}) , ignore!", root, node)
 			if err = r.client.Delete(zkPath); err == nil {
 				_, err = r.client.RegisterTemp(root, node)
 			}
 			if err != nil {
 				logger.Errorf("Recreate the temp node failed, (root{%s}, node{%s}) = error{%v}", root, node, perrors.WithStack(err))
-			}
+			}*/
```
Comment (Patrick0308) [inline registry/zookeeper/registry.go:237]: I don't know why delete these. It  may be make some problem.

File: config/reference_config.go [hunk 3]
```diff
@@ -257,9 +257,15 @@ func (c *ReferenceConfig) GenericLoad(id string) {
 	c.Implement(genericService)
 }
 
+// GetInvoker get invoker from ReferenceConfig
+func (c *ReferenceConfig) GetInvoker() protocol.Invoker {
+	return c.invoker
+}
+
```
[the remark this answers is not in the document]
Comment (jack15083) [reply config/reference_config.go:264]: https://github.com/apache/dubbo-go/pull/940 我是把这个pr也带进来了

File: registry/zookeeper/registry.go (hunk 2 above)
Comment (jack15083) [reply registry/zookeeper/registry.go:237]: zk连接丢失重连的时候，如果先删除再注册，会发出一个delete 的zkEnvent 跟 nodeChange zkEnvent 这两个事件消息是并行的，如果delete消息是后面处理的就会导致consumer cache  的invoker被删除，所以遇到这个报错应该ignore掉，认为是已经注册成功了

File: remoting/zookeeper/client.go (hunk 1 above)
Comment (jack15083) [reply remoting/zookeeper/client.go:269]: image
看这段代码有可能stop消息已经发出去了 但z.Conn还没清空，导致重连的时候失败，这段代码我是测试过了的，确实会出现这样的问题，虽然概率较低

File: registry/zookeeper/registry.go (hunk 2 above)
Comment (Patrick0308) [reply registry/zookeeper/registry.go:237]: Modifying the code here can't completely avoid the disorder of messages. We should try to recheck when we receive the event that the node is deleted. What do you think？@AlexStocks @pantianying @CodingSinger @zouyx

Comment (jack15083) [reply registry/zookeeper/registry.go:237]: 这段逻辑可能还不能删掉等我再测试一段时间再看吧

File: remoting/zookeeper/listener.go [hunk 4]
```diff
@@ -301,11 +299,8 @@ func (l *ZkEventListener) listenDirEvent(conf *common.URL, zkPath string, listen
 			go func(zkPath string, listener remoting.DataListener) {
 				if l.listenServiceNodeEvent(zkPath) {
 					listener.DataChange(remoting.Event{Path: zkPath, Action: remoting.EventTypeDel})
-					l.pathMapLock.Lock()
-					defer l.pathMapLock.Unlock()
```
Comment (wenxuwan) [inline remoting/zookeeper/listener.go:305]: why u remove the code? you can refer pr #850,

File: remoting/zookeeper/listener.go [hunk 5]
```diff
@@ -163,18 +167,17 @@ func (l *ZkEventListener) handleZkNodeEvent(zkPath string, children []string, li
...
 		go func(node string, zkPath string, listener remoting.DataListener) {
-			logger.Infof("delete zkNode{%s}", node)
 			if l.listenServiceNodeEvent(node, listener) {
-				logger.Infof("delete content{%s}", node)
-				listener.DataChange(remoting.Event{Path: zkPath, Action: remoting.EventTypeDel})
+				logger.Warnf("delete zkNode{%s}", node)
+				listener.DataChange(remoting.Event{Path: node, Action: remoting.EventTypeDel})
 			}
-			logger.Warnf("listenSelf(zk path{%s}) goroutine exit now", zkPath)
+			logger.Warnf("handleZkNodeEvent->listenSelf(zk path{%s}) goroutine exit now", zkPath)
```
Comment (wenxuwan) [inline remoting/zookeeper/listener.go:180]: zkPath is not needed for function, here is watching node not zkPath

File: remoting/zookeeper/listener.go [hunk 6]
```diff
@@ -113,6 +113,9 @@ func (l *ZkEventListener) listenServiceNodeEvent(zkPath string, listener ...remo
 			case zk.EventNotWatching:
 				logger.Warnf("zk.ExistW(key{%s}) = event{EventNotWatching}", zkPath)
 			case zk.EventNodeDeleted:
+				l.pathMapLock.Lock()
+				delete(l.pathMap, zkPath)
+				l.pathMapLock.Unlock()
```
Comment (wenxuwan) [inline remoting/zookeeper/listener.go:118]: if you put the delete logic here, ExistW on line82 return error will not delete the node.

File: remoting/zookeeper/client.go (hunk 1 above)
Comment (champly) [reply remoting/zookeeper/client.go:269]: 应该是stop发出去了，但z.Conn还没清空，重新`ValidateZookeeperClient`里面发现 Conn 不为空，不会重新连接(不会重连失败)，然后继续执行stop之后的逻辑就把连接关了

**Comment by champly** [comment]
我们也遇到了相同的问题，按照修改已经解决了

**Review by AlexStocks — APPROVED** [review]

File: remoting/zookeeper/listener.go (hunk 6 above)
Comment (jack15083) [reply remoting/zookeeper/listener.go:118]: 嗯明白了我再重新修改下

**Comment by cityiron** [comment]
You should resolve conflicts, @jack15083 hi, what your name in dingding?

**Comment by jack15083** [comment]
> You should resolve conflicts, @jack15083 hi, what your name in dingding?

冲突已经解决了

**Review by zouyx — APPROVED** [review]

File: config/reference_config.go [hunk 7]
```diff
@@ -269,4 +274,5 @@ func (c *ReferenceConfig) postProcessConfig(url *common.URL) {
 	for _, p := range extension.GetConfigPostProcessors() {
 		p.PostProcessReferenceConfig(url)
 	}
+
```
Comment (wenxuwan) [inline config/reference_config.go:277]: remove empty line.

File: remoting/zookeeper/facade.go [hunk 8]
```diff
@@ -82,6 +82,7 @@ LOOP:
 					break
 				}
 				failTimes++
+				logger.Warnf("ZK reconnect fail times %d", failTimes)
```
Comment (wenxuwan) [inline remoting/zookeeper/facade.go:85]: logger.Warnf("ZK reconnect failed %d times" , failTimes)

**Review by wenxuwan — APPROVED** [review]

**Pull request merged**